Node 0.11 compatibility?

I'm having trouble installing/compiling this for node 0.11.  Are there plans for supporting it?

FWIW, the errors I'm seeing are all basically this:

```
In file included from ../src/xattr.cc:9:
../src/error.cc:16:18: error: no member named 'New' in 'v8::String'
  return String::New(strerror(e));
```

I found https://github.com/rvagg/nan, which claims to provide native-code compatibility across many node versions.

Right now, I'm planning on trying to integrate that here and then opening a PR with the changes.  Any objections?

--- a/binding.gyp
+++ b/binding.gyp
@@ -2,7 +2,10 @@
   "targets": [
     {
       "target_name": "xattr",
-      "sources": [ "src/xattr.cc" ]
+      "sources": [ "src/xattr.cc" ],
+      "include_dirs" : [
+        "<!(node -e \"require('nan')\")"
+      ]
     }
   ]
 }

--- a/package.json
+++ b/package.json
@@ -14,5 +14,8 @@
   "repository": {
     "type": "git",
     "url": "http://github.com/LinusU/fs-xattr.git"
+  },
+  "dependencies": {
+    "nan": "^1.4.1"
   }
 }

--- a/src/error.cc
+++ b/src/error.cc
@@ -10,103 +10,101 @@
 #define S_ENOATTR "ENODATA"
 #endif
 
-using namespace v8;
-
-Local<String> errorDescription(int e) {
-  return String::New(strerror(e));
+const char* errorDescription(int e) {
+  return strerror(e);
 }
 
-Local<String> errorDescriptionForGet(int e) {
+const char* errorDescriptionForGet(int e) {
   switch (e) {
-    case E_ENOATTR: return String::New("The extended attribute does not exist.");
-    case ENOTSUP: return String::New("The file system does not support extended attributes or has the feature disabled.");
-    case ERANGE: return String::New("value (as indicated by size) is too small to hold the extended attribute data.");
-    case EPERM: return String::New("The named attribute is not permitted for this type of object.");
-    case EINVAL: return String::New("name is invalid or options has an unsupported bit set.");
-    case EISDIR: return String::New("path or fd do not refer to a regular file and the attribute in question is only applicable to files.  Similar to EPERM.");
-    case ENOTDIR: return String::New("A component of path's prefix is not a directory.");
-    case ENAMETOOLONG: return String::New("The length of name exceeds XATTR_MAXNAMELEN UTF-8 bytes, or a component of path exceeds NAME_MAX characters, or the entire path exceeds PATH_MAX characters.");
-    case EACCES: return String::New("Search permission is denied for a component of path or the attribute is not allowed to be read (e.g. an ACL prohibits reading the attributes of this file).");
-    case ELOOP: return String::New("Too many symbolic links were encountered in translating the pathname.");
-    case EFAULT: return String::New("path or name points to an invalid address.");
-    case EIO: return String::New("An I/O error occurred while reading from or writing to the file system.");
+    case E_ENOATTR: return "The extended attribute does not exist.";
+    case ENOTSUP: return "The file system does not support extended attributes or has the feature disabled.";
+    case ERANGE: return "value (as indicated by size) is too small to hold the extended attribute data.";
+    case EPERM: return "The named attribute is not permitted for this type of object.";
+    case EINVAL: return "name is invalid or options has an unsupported bit set.";
+    case EISDIR: return "path or fd do not refer to a regular file and the attribute in question is only applicable to files.  Similar to EPERM.";
+    case ENOTDIR: return "A component of path's prefix is not a directory.";
+    case ENAMETOOLONG: return "The length of name exceeds XATTR_MAXNAMELEN UTF-8 bytes, or a component of path exceeds NAME_MAX characters, or the entire path exceeds PATH_MAX characters.";
+    case EACCES: return "Search permission is denied for a component of path or the attribute is not allowed to be read (e.g. an ACL prohibits reading the attributes of this file).";
+    case ELOOP: return "Too many symbolic links were encountered in translating the pathname.";
+    case EFAULT: return "path or name points to an invalid address.";
+    case EIO: return "An I/O error occurred while reading from or writing to the file system.";
     default: return errorDescription(e);
   }
 }
 
-Local<String> errorDescriptionForSet(int e) {
+const char* errorDescriptionForSet(int e) {
   switch (e) {
-    case EEXIST: return String::New("options contains XATTR_CREATE and the named attribute already exists.");
-    case E_ENOATTR: return String::New("options is set to XATTR_REPLACE and the named attribute does not exist.");
-    case ENOTSUP: return String::New("The file system does not support extended attributes or has the feature disabled.");
-    case EROFS: return String::New("The file system is mounted read-only.");
-    case ERANGE: return String::New("The data size of the attribute is out of range (some attributes have size restrictions).");
-    case EPERM: return String::New("Attributes cannot be associated with this type of object. For example, attributes are not allowed for resource forks.");
-    case EINVAL: return String::New("name or options is invalid. name must be valid UTF-8 and options must make sense.");
-    case ENOTDIR: return String::New("A component of path is not a directory.");
-    case ENAMETOOLONG: return String::New("name exceeded XATTR_MAXNAMELEN UTF-8 bytes, or a component of path exceeded NAME_MAX characters, or the entire path exceeded PATH_MAX characters.");
-    case EACCES: return String::New("Search permission is denied for a component of path or permission to set the attribute is denied.");
-    case ELOOP: return String::New("Too many symbolic links were encountered resolving path.");
-    case EFAULT: return String::New("path or name points to an invalid address.");
-    case EIO: return String::New("An I/O error occurred while reading from or writing to the file system.");
-    case E2BIG: return String::New("The data size of the extended attribute is too large.");
-    case ENOSPC: return String::New("Not enough space left on the file system.");
+    case EEXIST: return "options contains XATTR_CREATE and the named attribute already exists.";
+    case E_ENOATTR: return "options is set to XATTR_REPLACE and the named attribute does not exist.";
+    case ENOTSUP: return "The file system does not support extended attributes or has the feature disabled.";
+    case EROFS: return "The file system is mounted read-only.";
+    case ERANGE: return "The data size of the attribute is out of range (some attributes have size restrictions).";
+    case EPERM: return "Attributes cannot be associated with this type of object. For example, attributes are not allowed for resource forks.";
+    case EINVAL: return "name or options is invalid. name must be valid UTF-8 and options must make sense.";
+    case ENOTDIR: return "A component of path is not a directory.";
+    case ENAMETOOLONG: return "name exceeded XATTR_MAXNAMELEN UTF-8 bytes, or a component of path exceeded NAME_MAX characters, or the entire path exceeded PATH_MAX characters.";
+    case EACCES: return "Search permission is denied for a component of path or permission to set the attribute is denied.";
+    case ELOOP: return "Too many symbolic links were encountered resolving path.";
+    case EFAULT: return "path or name points to an invalid address.";
+    case EIO: return "An I/O error occurred while reading from or writing to the file system.";
+    case E2BIG: return "The data size of the extended attribute is too large.";
+    case ENOSPC: return "Not enough space left on the file system.";
     default: return errorDescription(e);
   }
 }
 
-Local<String> errorCode(int e) {
+const char* errorCode(int e) {
   switch (e) {
     /* Standard */
-    case EPERM: return String::New("EPERM");
-    case ENOENT: return String::New("ENOENT");
-    case ESRCH: return String::New("ESRCH");
-    case EINTR: return String::New("EINTR");
-    case EIO: return String::New("EIO");
-    case ENXIO: return String::New("ENXIO");
-    case E2BIG: return String::New("E2BIG");
-    case ENOEXEC: return String::New("ENOEXEC");
-    case EBADF: return String::New("EBADF");
-    case ECHILD: return String::New("ECHILD");
-    case EAGAIN: return String::New("EAGAIN");
-    case ENOMEM: return String::New("ENOMEM");
-    case EACCES: return String::New("EACCES");
-    case EFAULT: return String::New("EFAULT");
-    case ENOTBLK: return String::New("ENOTBLK");
-    case EBUSY: return String::New("EBUSY");
-    case EEXIST: return String::New("EEXIST");
-    case EXDEV: return String::New("EXDEV");
-    case ENODEV: return String::New("ENODEV");
-    case ENOTDIR: return String::New("ENOTDIR");
-    case EISDIR: return String::New("EISDIR");
-    case EINVAL: return String::New("EINVAL");
-    case ENFILE: return String::New("ENFILE");
-    case EMFILE: return String::New("EMFILE");
-    case ENOTTY: return String::New("ENOTTY");
-    case ETXTBSY: return String::New("ETXTBSY");
-    case EFBIG: return String::New("EFBIG");
-    case ENOSPC: return String::New("ENOSPC");
-    case ESPIPE: return String::New("ESPIPE");
-    case EROFS: return String::New("EROFS");
-    case EMLINK: return String::New("EMLINK");
-    case EPIPE: return String::New("EPIPE");
-    case EDOM: return String::New("EDOM");
-    case ERANGE: return String::New("ERANGE");
+    case EPERM: return "EPERM";
+    case ENOENT: return "ENOENT";
+    case ESRCH: return "ESRCH";
+    case EINTR: return "EINTR";
+    case EIO: return "EIO";
+    case ENXIO: return "ENXIO";
+    case E2BIG: return "E2BIG";
+    case ENOEXEC: return "ENOEXEC";
+    case EBADF: return "EBADF";
+    case ECHILD: return "ECHILD";
+    case EAGAIN: return "EAGAIN";
+    case ENOMEM: return "ENOMEM";
+    case EACCES: return "EACCES";
+    case EFAULT: return "EFAULT";
+    case ENOTBLK: return "ENOTBLK";
+    case EBUSY: return "EBUSY";
+    case EEXIST: return "EEXIST";
+    case EXDEV: return "EXDEV";
+    case ENODEV: return "ENODEV";
+    case ENOTDIR: return "ENOTDIR";
+    case EISDIR: return "EISDIR";
+    case EINVAL: return "EINVAL";
+    case ENFILE: return "ENFILE";
+    case EMFILE: return "EMFILE";
+    case ENOTTY: return "ENOTTY";
+    case ETXTBSY: return "ETXTBSY";
+    case EFBIG: return "EFBIG";
+    case ENOSPC: return "ENOSPC";
+    case ESPIPE: return "ESPIPE";
+    case EROFS: return "EROFS";
+    case EMLINK: return "EMLINK";
+    case EPIPE: return "EPIPE";
+    case EDOM: return "EDOM";
+    case ERANGE: return "ERANGE";
     /* Special */
-    case E_ENOATTR: return String::New(S_ENOATTR);
-    case ENOTSUP: return String::New("ENOTSUP");
-    case ENAMETOOLONG: return String::New("ENAMETOOLONG");
-    case ELOOP: return String::New("ELOOP");
+    case E_ENOATTR: return S_ENOATTR;
+    case ENOTSUP: return "ENOTSUP";
+    case ENAMETOOLONG: return "ENAMETOOLONG";
+    case ELOOP: return "ELOOP";
     /* Unknown */
-    default: return String::New("");
+    default: return "";
   }
 }
 
 void ThrowExceptionErrno(int e) {
 
-  Local<Object> err = v8::Exception::Error(errorDescriptionForGet(e))->ToObject();
-  err->Set(String::NewSymbol("errno"), Integer::New(e));
-  err->Set(String::NewSymbol("code"), errorCode(e));
+  Local<Object> err = v8::Exception::Error(NanNew(errorDescriptionForGet(e)))->ToObject();
+  err->Set(NanNew("errno"), NanNew<Integer>(e));
+  err->Set(NanNew("code"), NanNew(errorCode(e)));
 
-  ThrowException(err);
+  NanThrowError(err);
 }

--- a/src/xattr.cc
+++ b/src/xattr.cc
@@ -1,20 +1,30 @@
 
-#include <v8.h>
 #include <node.h>
+#include <nan.h>
 #include <node_buffer.h>
 
 #include <stdlib.h>
 #include <sys/xattr.h>
 
+using v8::FunctionTemplate;
+using v8::Handle;
+using v8::HandleScope;
+using v8::String;
+
+using v8::Local;
+using v8::Object;
+using v8::Integer;
+using v8::Value;
+using v8::Function;
+using v8::Array;
+
 #include "error.cc"
 
-using namespace v8;
+NAN_METHOD(xattr_get) {
+  NanScope();
 
-Handle<Value> xattr_get(const Arguments& args) {
-  HandleScope scope;
-
-  String::AsciiValue aFilename(args[0]);
-  String::AsciiValue aAttribute(args[1]);
+  NanAsciiString aFilename(args[0]);
+  NanAsciiString aAttribute(args[1]);
 
   const char *filename = *aFilename;
   const char *attribute = *aAttribute;
@@ -29,10 +39,10 @@ Handle<Value> xattr_get(const Arguments& args) {
 
   if (valueLen == -1) {
     ThrowExceptionErrno(errno);
-    return scope.Close(Undefined());
+    NanReturnUndefined();
   }
 
-  node::Buffer *slowBuffer = node::Buffer::New((size_t) valueLen);
+  Local<Object> slowBuffer = NanNewBufferHandle((size_t) valueLen);
 
 #ifdef __APPLE__
   valueLen = getxattr(filename, attribute, node::Buffer::Data(slowBuffer), (size_t) valueLen, 0, 0);
@@ -42,22 +52,22 @@ Handle<Value> xattr_get(const Arguments& args) {
 
   if (valueLen == -1) {
     ThrowExceptionErrno(errno);
-    return scope.Close(Undefined());
+    NanReturnUndefined();
   }
 
-  Local<Object> globalObj = Context::GetCurrent()->Global();
-  Local<Function> bufferConstructor = Local<Function>::Cast(globalObj->Get(String::NewSymbol("Buffer")));
-  Handle<Value> constructorArgs[3] = { slowBuffer->handle_, Integer::New(valueLen), Integer::New(0) };
+  Local<Object> globalObj = NanGetCurrentContext()->Global();
+  Local<Function> bufferConstructor = Local<Function>::Cast(globalObj->Get(NanNew("Buffer")));
+  Handle<Value> constructorArgs[3] = { slowBuffer, NanNew<Integer>(valueLen), NanNew<Integer>(0) };
   Local<Object> actualBuffer = bufferConstructor->NewInstance(3, constructorArgs);
 
-  return scope.Close(actualBuffer);
+  NanReturnValue(actualBuffer);
 }
 
-Handle<Value> xattr_set(const Arguments& args) {
-  HandleScope scope;
+NAN_METHOD(xattr_set) {
+  NanScope();
 
-  String::AsciiValue aFilename(args[0]);
-  String::AsciiValue aAttribute(args[1]);
+  NanAsciiString aFilename(args[0]);
+  NanAsciiString aAttribute(args[1]);
 
   const char *filename = *aFilename;
   const char *attribute = *aAttribute;
@@ -74,16 +84,16 @@ Handle<Value> xattr_set(const Arguments& args) {
 
   if (res == -1) {
     ThrowExceptionErrno(errno);
-    return scope.Close(Undefined());
+    NanReturnUndefined();
   }
 
-  return scope.Close(Undefined());
+  NanReturnUndefined();
 }
 
-Handle<Value> xattr_list(const Arguments& args) {
-  HandleScope scope;
+NAN_METHOD(xattr_list) {
+  NanScope();
 
-  String::AsciiValue aFilename(args[0]);
+  NanAsciiString aFilename(args[0]);
   const char *filename = *aFilename;
 
   ssize_t valueLen;
@@ -96,7 +106,7 @@ Handle<Value> xattr_list(const Arguments& args) {
 
   if (valueLen == -1) {
     ThrowExceptionErrno(errno);
-    return scope.Close(Undefined());
+    NanReturnUndefined();
   }
 
   char *result = (char *) malloc((size_t) valueLen);
@@ -109,7 +119,7 @@ Handle<Value> xattr_list(const Arguments& args) {
 
   if (valueLen == -1) {
     ThrowExceptionErrno(errno);
-    return scope.Close(Undefined());
+    NanReturnUndefined();
   }
 
   Local<Array> arr = Array::New(0);
@@ -118,22 +128,22 @@ Handle<Value> xattr_list(const Arguments& args) {
   ssize_t valuePos = 0;
   while (valuePos < valueLen) {
     size_t len = strlen(result);
-    arr->Set(arrayPos, String::New(result, len));
+    arr->Set(arrayPos, NanNew<String>(result, len));
     valuePos += len + 1;
     arrayPos++;
   }
 
-  arr->Set(String::NewSymbol("length"), Integer::New(arrayPos));
+  arr->Set(NanNew("length"), NanNew<Integer>(arrayPos));
   free(result);
 
-  return scope.Close(arr);
+  NanReturnValue(arr);
 }
 
-Handle<Value> xattr_remove(const Arguments& args) {
-  HandleScope scope;
+NAN_METHOD(xattr_remove) {
+  NanScope();
 
-  String::AsciiValue aFilename(args[0]);
-  String::AsciiValue aAttribute(args[1]);
+  NanAsciiString aFilename(args[0]);
+  NanAsciiString aAttribute(args[1]);
 
   const char *filename = *aFilename;
   const char *attribute = *aAttribute;
@@ -146,17 +156,17 @@ Handle<Value> xattr_remove(const Arguments& args) {
 
   if (res == -1) {
     ThrowExceptionErrno(errno);
-    return scope.Close(Undefined());
+    NanReturnUndefined();
   }
 
-  return scope.Close(Undefined());
+  NanReturnUndefined();
 }
 
 void Initialize(Handle<Object> exports) {
-  exports->Set(String::NewSymbol("get"), FunctionTemplate::New(xattr_get)->GetFunction());
-  exports->Set(String::NewSymbol("set"), FunctionTemplate::New(xattr_set)->GetFunction());
-  exports->Set(String::NewSymbol("list"), FunctionTemplate::New(xattr_list)->GetFunction());
-  exports->Set(String::NewSymbol("remove"), FunctionTemplate::New(xattr_remove)->GetFunction());
+  exports->Set(NanNew("get"), NanNew<FunctionTemplate>(xattr_get)->GetFunction());
+  exports->Set(NanNew("set"), NanNew<FunctionTemplate>(xattr_set)->GetFunction());
+  exports->Set(NanNew("list"), NanNew<FunctionTemplate>(xattr_list)->GetFunction());
+  exports->Set(NanNew("remove"), NanNew<FunctionTemplate>(xattr_remove)->GetFunction());
 }
 
 NODE_MODULE(xattr, Initialize)

